prow/gcs/reporter/reporter_test.go

```diff
@@ -0,0 +1,185 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reporter
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/GoogleCloudPlatform/testgrid/metadata"
+	"github.com/fsouza/fake-gcs-server/fakestorage"
```

Comment:
Thoughts about leveraging an interface here?

You depend on three methods in the storage client:

```golang
w := storageClient.Bucket(b).Object(path).If().NewWriter()
w.Write()
w.Close()
```

which can be abstracted into the following interface that returns a built-in interface:

```golang
type author interface {
  NewWriter(context.Context, path gcs.Path) io.WriterCloser // or NewWriter(bucket, path string)
}


type storageAuthor struct {
  client *storage.Client
}

func (sa storageAuthor) NewWriter(ctx context.Context, path gcs.Path) io.WriterCloser {
  return sa.client.Bucket(path.Bucket()).Object(path.Object()).If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
}
```

This should obviate the need to use the fakestorage import and also make it easier to write smaller and more hermetic unit tests.

Reply:
I did this.